// @enact/moonstone/FormCheckboxItem/FormCheckboxItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FormCheckboxItemBase = exports.FormCheckboxItem = exports["default"] = void 0;

var _kind = _interopRequireDefault(require("@enact/core/kind"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _FormCheckbox = _interopRequireDefault(require("../FormCheckbox"));

var _ToggleItem = _interopRequireDefault(require("../ToggleItem"));

var _FormCheckboxItemModule = _interopRequireDefault(require("./FormCheckboxItem.module.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

/**
 * Provides Moonstone styled form item component and interactive toggleable checkbox.
 *
 * @example
 * <FormCheckboxItem>A Checkbox for a form</FormCheckboxItem>
 *
 * @module moonstone/FormCheckboxItem
 * @exports FormCheckboxItem
 * @exports FormCheckboxItemBase
 */

/**
 * Renders a form item with a checkbox component. Useful to show a selected state on an item inside a form.
 *
 * @class FormCheckboxItem
 * @memberof moonstone/FormCheckboxItem
 * @extends moonstone/ToggleItem.ToggleItem
 * @omit iconComponent
 * @ui
 * @public
 */
var FormCheckboxItemBase = (0, _kind["default"])({
  name: 'FormCheckboxItem',
  propTypes:
  /** @lends moonstone/FormCheckboxItem.FormCheckboxItem.prototype */
  {
    /**
     * Customizes the component by mapping the supplied collection of CSS class names to the
     * corresponding internal Elements and states of this component.
     *
     * The following classes are supported:
     *
     * * `formCheckboxItem` - The root class name
     *
     * @type {Object}
     * @public
     */
    css: _propTypes["default"].object
  },
  styles: {
    css: _FormCheckboxItemModule["default"],
    className: 'formCheckboxItem',
    publicClassNames: ['formCheckboxItem']
  },
  render: function render(props) {
    return _react["default"].createElement(_ToggleItem["default"], Object.assign({
      "data-webos-voice-intent": "SelectCheckItem"
    }, props, {
      css: props.css,
      iconComponent: _FormCheckbox["default"]
    }));
  }
});
exports.FormCheckboxItemBase = exports.FormCheckboxItem = FormCheckboxItemBase;
var _default = FormCheckboxItemBase;
exports["default"] = _default;
