// @enact/ui/ViewManager/arrange.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.compose = exports.ease = exports.endBy = exports.startAfter = exports.reverse = exports.slideOut = exports.slideIn = exports.slideOutPartial = exports.slideInPartial = exports.accelerate = exports.fadeIn = exports.fadeOut = exports.clearTransform = exports.replaceTransform = exports.prependTransform = exports.appendTransform = exports.transform = void 0;

var _curry = _interopRequireDefault(require("ramda/src/curry"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var transform = (0, _curry["default"])(function (action, spec, _ref) {
  var node = _ref.node;
  var current = node.style.transform;
  var next = current;

  if (!current || action === 'replace') {
    next = spec;
  } else if (action === 'append') {
    next = current + ' ' + spec;
  } else if (action === 'prepend') {
    next = spec + ' ' + current;
  }

  node.style.transform = next;
});
exports.transform = transform;
var appendTransform = transform('append');
exports.appendTransform = appendTransform;
var prependTransform = transform('prepend');
exports.prependTransform = prependTransform;
var replaceTransform = transform('replace');
exports.replaceTransform = replaceTransform;
var clearTransform = transform('replace', '');
exports.clearTransform = clearTransform;

var fadeOut = function fadeOut(_ref2) {
  var node = _ref2.node,
      percent = _ref2.percent;
  node.style.opacity = 1 - percent;
};

exports.fadeOut = fadeOut;

var fadeIn = function fadeIn(_ref3) {
  var node = _ref3.node,
      percent = _ref3.percent;
  node.style.opacity = percent;
};

exports.fadeIn = fadeIn;
var accelerate = prependTransform('translateZ(0)');
exports.accelerate = accelerate;
var slideInOut = (0, _curry["default"])(function (direction, total, orientation, config) {
  var percent = config.percent;
  var p = total * (direction === 'out' ? percent : 1 - percent);
  var spec = orientation === 'top' && 'translateY(' + -p + '%)' || orientation === 'bottom' && 'translateY(' + p + '%)' || orientation === 'left' && 'translateX(' + -p + '%)' || orientation === 'right' && 'translateX(' + p + '%)';
  appendTransform(spec, config);
});
var slideInPartial = slideInOut('in');
exports.slideInPartial = slideInPartial;
var slideOutPartial = slideInOut('out');
exports.slideOutPartial = slideOutPartial;
var slideIn = slideInPartial(100);
exports.slideIn = slideIn;
var slideOut = slideOutPartial(100);
exports.slideOut = slideOut;
var reverse = (0, _curry["default"])(function (fn, _ref4) {
  var reverseTransition = _ref4.reverseTransition,
      percent = _ref4.percent,
      rest = _objectWithoutProperties(_ref4, ["reverseTransition", "percent"]);

  fn(_objectSpread({}, rest, {
    percent: reverseTransition ? 1 - percent : percent,
    reverseTransition: reverseTransition
  }));
});
exports.reverse = reverse;
var startAfter = (0, _curry["default"])(function (startPercent, fn, config) {
  var percent = config.percent,
      rest = _objectWithoutProperties(config, ["percent"]);

  var p = percent >= startPercent ? (percent - startPercent) / (1 - startPercent) : 0;
  fn(_objectSpread({}, rest, {
    percent: p
  }));
});
exports.startAfter = startAfter;
var endBy = (0, _curry["default"])(function (endPercent, fn, config) {
  var percent = config.percent,
      rest = _objectWithoutProperties(config, ["percent"]);

  var p = percent <= endPercent ? percent / endPercent : 1;
  fn(_objectSpread({}, rest, {
    percent: p
  }));
});
exports.endBy = endBy;
var ease = (0, _curry["default"])(function (easing, fn, config) {
  var percent = config.percent,
      rest = _objectWithoutProperties(config, ["percent"]);

  fn(_objectSpread({}, rest, {
    percent: easing(percent)
  }));
});
exports.ease = ease;

var compose = function compose() {
  for (var _len = arguments.length, arrangers = new Array(_len), _key = 0; _key < _len; _key++) {
    arrangers[_key] = arguments[_key];
  }

  return function () {
    for (var _len2 = arguments.length, args = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
      args[_key2] = arguments[_key2];
    }

    arrangers.forEach(function (fn) {
      return fn.apply(void 0, args);
    });
  };
};

exports.compose = compose;
