// @enact/ui/Placeholder/PlaceholderDecorator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PlaceholderDecorator = exports["default"] = void 0;

var _hoc = _interopRequireDefault(require("@enact/core/hoc"));

var _react = _interopRequireDefault(require("react"));

var _PlaceholderControllerDecorator = require("./PlaceholderControllerDecorator");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

/**
 * Default config for PlaceholderDecorator.
 *
 * @memberof ui/Placeholder.PlaceholderDecorator
 * @hocconfig
 * @public
 */
var defaultConfig = {
  /**
   * Configures the style of the placeholder element
   *
   * @type {Object}
   * @default {height: 0, width: 'auto'}
   * @memberof ui/Placeholder.PlaceholderDecorator.defaultConfig
   */
  style: {
    height: 0,
    width: 'auto'
  },

  /**
   * The component to use as a placeholder.
   *
   * @type {String}
   * @default 'div'
   * @memberof ui/Placeholder.PlaceholderDecorator.defaultConfig
   */
  placeholderComponent: 'div'
};
/**
 * A higher-order component that enables a container to notify the wrapped component when scrolling.
 *
 * Containers must provide `registerPlaceholder`, `unregisterPlaceholder`, and `invalidateBounds`
 * methods via React's context for `PlaceholderDecorator` instances.
 *
 * @class PlaceholderDecorator
 * @memberof ui/Placeholder
 * @hoc
 * @public
 */

var PlaceholderDecorator = (0, _hoc["default"])(defaultConfig, function (config, Wrapped) {
  var _class, _temp;

  var PlaceholderComponent = config.placeholderComponent,
      style = config.style;
  var placeholderStyle = Object.assign({}, defaultConfig.style, style);
  return _temp = _class =
  /*#__PURE__*/
  function (_React$PureComponent) {
    _inherits(_class, _React$PureComponent);

    function _class() {
      var _this;

      _classCallCheck(this, _class);

      _this = _possibleConstructorReturn(this, _getPrototypeOf(_class).call(this));

      _this.initPlaceholderRef = function (ref) {
        _this.placeholderRef = ref;
      };

      _this.state = {
        visible: false
      };
      return _this;
    }

    _createClass(_class, [{
      key: "componentDidMount",
      value: function componentDidMount() {
        if (!this.state.visible && this.context) {
          this.controller = this.context(this.update.bind(this));
        }
      }
    }, {
      key: "componentDidUpdate",
      value: function componentDidUpdate(prevProps, prevState) {
        if (this.controller && this.state.visible && prevState.visible !== this.state.visible) {
          // also need invalidateBounds() for scroller :(
          this.controller.unregister();
        }
      }
    }, {
      key: "componentWillUnmount",
      value: function componentWillUnmount() {
        if (this.controller && !this.state.visible) {
          this.controller.unregister();
        }
      }
    }, {
      key: "update",
      value: function update(_ref) {
        var leftThreshold = _ref.leftThreshold,
            topThreshold = _ref.topThreshold;
        var _this$placeholderRef = this.placeholderRef,
            offsetLeft = _this$placeholderRef.offsetLeft,
            offsetTop = _this$placeholderRef.offsetTop,
            offsetHeight = _this$placeholderRef.offsetHeight,
            offsetWidth = _this$placeholderRef.offsetWidth;

        if (offsetTop < topThreshold + offsetHeight && offsetLeft < leftThreshold + offsetWidth) {
          this.setState(function (state) {
            return state.visible ? null : {
              visible: true
            };
          });
        }
      }
    }, {
      key: "render",
      value: function render() {
        var visible = this.state.visible;

        if (visible) {
          return _react["default"].createElement(Wrapped, Object.assign({}, this.props, {
            ref: this.initPlaceholderRef
          }));
        } else {
          return _react["default"].createElement(PlaceholderComponent, {
            ref: this.initPlaceholderRef,
            style: placeholderStyle
          });
        }
      }
    }]);

    return _class;
  }(_react["default"].PureComponent), _class.displayName = 'PlaceholderDecorator', _class.contextType = _PlaceholderControllerDecorator.PlaceholderContext, _temp;
});
exports.PlaceholderDecorator = PlaceholderDecorator;
var _default = PlaceholderDecorator;
exports["default"] = _default;
