// @enact/ui/Button/Button.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ButtonDecorator = exports.ButtonBase = exports.Button = exports["default"] = void 0;

var _kind = _interopRequireDefault(require("@enact/core/kind"));

var _propTypes = _interopRequireDefault(require("@enact/core/internal/prop-types"));

var _propTypes2 = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _Touchable = _interopRequireDefault(require("../Touchable"));

var _ButtonModule = _interopRequireDefault(require("./Button.module.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/**
 * A basic button component structure without any behaviors applied to it.
 *
 * @class ButtonBase
 * @memberof ui/Button
 * @ui
 * @public
 */
var ButtonBase = (0, _kind["default"])({
  name: 'ui:Button',
  propTypes:
  /** @lends ui/Button.ButtonBase.prototype */
  {
    children: _propTypes2["default"].node.isRequired,

    /**
     * Customizes the component by mapping the supplied collection of CSS class names to the
     * corresponding internal Elements and states of this component.
     *
     * The following classes are supported:
     *
     * * `button` - The root component class
     * * `bg` - The background node of the button
     * * `client` - The content node of the button
     * * `icon` - The icon node, when `icon` is set
     * * `minWidth` - Applied when `minWidth` prop is `true`
     * * `pressed` - Applied when `pressed` prop is `true`
     * * `selected` - Applied when `selected` prop is `true`
     * * `small` - Applied when `small` prop is `true`
     *
     * @type {Object}
     * @public
     */
    css: _propTypes2["default"].object,

    /**
     * Applies the `disabled` class.
     *
     * When `true`, the button is shown as disabled.
     *
     * @type {Boolean}
     * @default false
     * @public
     */
    disabled: _propTypes2["default"].bool,

    /**
     * The icon displayed within the Button.
     *
     * The icon will be displayed before the natural reading order of the text, regardless
     * of locale. Any string that is valid for its {@link ui/Button.Button.iconComponent} is
     * valid here. If `icon` is specified as a string and `iconButton` is undefined, the icon is
     * not rendered.
     *
     * This also supports a custom icon, in the form of a DOM node or a Component,
     * with the caveat that if you supply a custom icon, you are responsible for sizing and
     * locale positioning of the custom component.
     *
     * @type {Node}
     * @public
     */
    icon: _propTypes2["default"].node,

    /**
     * The component used to render the [icon]{@link ui/Button.ButtonBase.icon}.
     *
     * This component will receive the `small` property set on the Button as well as the `icon`
     * class to customize its styling. If [icon]{@link ui/Button.ButtonBase.icon} is not a
     * string, this property is not used.
     *
     * @type {Component}
     * @public
     */
    iconComponent: _propTypes["default"].component,

    /**
     * Enforces a minimum width for the component.
     *
     * Applies the `minWidth` CSS class which can be customized by
     * [theming]{@link /docs/developer-guide/theming/}.
     *
     * @type {Boolean}
     * @default true
     * @public
     */
    minWidth: _propTypes2["default"].bool,

    /**
     * Indicates the component is depressed.
     *
     * Applies the `pressed` CSS class which can be customized by
     * [theming]{@link /docs/developer-guide/theming/}.
     *
     * @type {Boolean}
     * @default false
     * @public
     */
    pressed: _propTypes2["default"].bool,

    /**
     * Indicates the component is selected.
     *
     * Applies the `selected` CSS class which can be customized by
     * [theming]{@link /docs/developer-guide/theming/}.
     *
     * @type {Boolean}
     * @default false
     * @public
     */
    selected: _propTypes2["default"].bool,

    /**
     * Reduces the size of the component.
     *
     * Applies the `small` CSS class which can be customized by
     * [theming]{@link /docs/developer-guide/theming/}.
     *
     * @type {Boolean}
     * @default false
     * @public
     */
    small: _propTypes2["default"].bool
  },
  defaultProps: {
    disabled: false,
    minWidth: true,
    pressed: false,
    selected: false,
    small: false
  },
  styles: {
    css: _ButtonModule["default"],
    className: 'button',
    publicClassNames: true
  },
  computed: {
    className: function className(_ref) {
      var minWidth = _ref.minWidth,
          pressed = _ref.pressed,
          selected = _ref.selected,
          small = _ref.small,
          styler = _ref.styler;
      return styler.append({
        pressed: pressed,
        small: small,
        minWidth: minWidth,
        selected: selected
      });
    },
    icon: function icon(_ref2) {
      var css = _ref2.css,
          _icon = _ref2.icon,
          Icon = _ref2.iconComponent,
          small = _ref2.small;
      return typeof _icon === 'string' && Icon ? _react["default"].createElement(Icon, {
        small: small,
        className: css.icon
      }, _icon) : _icon;
    }
  },
  render: function render(_ref3) {
    var children = _ref3.children,
        css = _ref3.css,
        disabled = _ref3.disabled,
        icon = _ref3.icon,
        rest = _objectWithoutProperties(_ref3, ["children", "css", "disabled", "icon"]);

    delete rest.iconComponent;
    delete rest.minWidth;
    delete rest.pressed;
    delete rest.selected;
    delete rest.small;
    return _react["default"].createElement("div", Object.assign({
      role: "button"
    }, rest, {
      "aria-disabled": disabled,
      disabled: disabled
    }), _react["default"].createElement("div", {
      className: css.bg
    }), _react["default"].createElement("div", {
      className: css.client
    }, icon, children));
  }
});
/**
 * A higher-order component that adds touch support to a [ButtonBase]{@link ui/Button.ButtonBase}.
 *
 * @hoc
 * @memberof ui/Button
 * @mixes ui/Touchable.Touchable
 * @public
 */

exports.ButtonBase = ButtonBase;
var ButtonDecorator = (0, _Touchable["default"])({
  activeProp: 'pressed'
});
/**
 * A minimally-styled button component with touch support.
 *
 * @class Button
 * @extends ui/Button.ButtonBase
 * @memberof ui/Button
 * @mixes ui/Button.ButtonDecorator
 * @ui
 * @public
 */

exports.ButtonDecorator = ButtonDecorator;
var Button = ButtonDecorator(ButtonBase);
exports.Button = Button;
var _default = Button;
exports["default"] = _default;
