// @enact/ui/GridListImageItem/GridListImageItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GridListImageItem = exports["default"] = void 0;

var _kind = _interopRequireDefault(require("@enact/core/kind"));

var _propTypes = _interopRequireDefault(require("@enact/core/internal/prop-types"));

var _Layout = require("@enact/ui/Layout");

var _propTypes2 = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _Icon = _interopRequireDefault(require("../Icon"));

var _Image = _interopRequireDefault(require("../Image"));

var _GridListImageItemModule = _interopRequireDefault(require("./GridListImageItem.module.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/**
 * A basic grid list image item without any behavior.
 *
 * @class GridListImageItem
 * @memberof ui/GridListImageItem
 * @ui
 * @public
 */
var GridListImageItem = (0, _kind["default"])({
  name: 'ui:GridListImageItem',
  propTypes:
  /** @lends ui/GridListImageItem.GridListImageItem.prototype */
  {
    /**
     * The primary caption to be displayed with the image.
     *
     * @type {String}
     * @public
     */
    caption: _propTypes2["default"].string,

    /**
     * The component used to render the captions
     *
     * @type {String|Component}
     * @public
     */
    captionComponent: _propTypes["default"].renderable,

    /**
     * Customizes the component by mapping the supplied collection of CSS class names to the
     * corresponding internal Elements and states of this component.
     *
     * The following classes are supported:
     *
     * * `icon` - The icon component class for default selection overlay
     * * `image` - The image component class
     * * `selected` - Applied when `selected` prop is `true`
     * * `caption` - The caption component class
     * * `subCaption` - The subCaption component class
     *
     * @type {Object}
     * @public
     */
    css: _propTypes2["default"].object,

    /**
     * The component used to render the default check icon when selected.
     * If there is custom selectionOverlay component, this icon will not be shown.
     *
     * @type {Component}
     * @default ui/Icon.Icon
     * @public
     */
    iconComponent: _propTypes["default"].component,

    /**
     * The component used to render the image component
     *
     * @type {Component}
     * @default ui/Image.Image
     * @public
     */
    imageComponent: _propTypes["default"].component,

    /**
     * Placeholder image used while [source]{@link ui/GridListImageItem.GridListImageItem#source}
     * is loaded.
     *
     * @type {String}
     * @public
     */
    placeholder: _propTypes2["default"].string,

    /**
     * Applies a selected visual effect to the image, but only if `selectionOverlayShowing`
     * is also `true`.
     *
     * @type {Boolean}
     * @default false
     * @public
     */
    selected: _propTypes2["default"].bool,

    /**
     * The custom selection overlay component to render.
     *
     * A component can be a stateless functional component, `kind()` or React component.
     * The following is an example with custom selection overlay kind.
     *
     * Example:
     * ```
     * const SelectionOverlay = kind({
     * 	render: () => <div>custom overlay</div>
     * });
     *
     * <GridListImageItem selectionOverlay={SelectionOverlay} />
     * ```
     *
     * @type {Function}
     * @public
     */
    selectionOverlay: _propTypes2["default"].func,

    /**
     * Shows a selection overlay with a centered icon. When `selected` is true, a check mark is shown.
     *
     * @type {Boolean}
     * @default false
     * @public
     */
    selectionOverlayShowing: _propTypes2["default"].bool,

    /**
     * The absolute URL path to the image.
     *
     * @type {String}
     * @public
     */
    source: _propTypes2["default"].string,

    /**
     * The second caption line to be displayed with the image.
     *
     * @type {String}
     * @public
     */
    subCaption: _propTypes2["default"].string
  },
  defaultProps: {
    captionComponent: 'div',
    iconComponent: _Icon["default"],
    imageComponent: _Image["default"],
    selected: false,
    selectionOverlayShowing: false
  },
  styles: {
    css: _GridListImageItemModule["default"],
    className: 'gridListImageItem',
    publicClassNames: true
  },
  computed: {
    className: function className(_ref) {
      var caption = _ref.caption,
          selected = _ref.selected,
          styler = _ref.styler,
          subCaption = _ref.subCaption;
      return styler.append({
        selected: selected,
        caption: caption,
        subCaption: subCaption
      });
    },
    selectionOverlay: function selectionOverlay(_ref2) {
      var css = _ref2.css,
          IconComponent = _ref2.iconComponent,
          SelectionOverlay = _ref2.selectionOverlay,
          selectionOverlayShowing = _ref2.selectionOverlayShowing;

      if (selectionOverlayShowing) {
        return _react["default"].createElement("div", {
          className: css.overlayContainer
        }, SelectionOverlay ? _react["default"].createElement(SelectionOverlay, null) : _react["default"].createElement("div", {
          className: css.overlayComponent
        }, _react["default"].createElement(IconComponent, {
          className: css.icon
        }, "check")));
      }
    }
  },
  render: function render(_ref3) {
    var caption = _ref3.caption,
        Caption = _ref3.captionComponent,
        css = _ref3.css,
        ImageComponent = _ref3.imageComponent,
        placeholder = _ref3.placeholder,
        source = _ref3.source,
        subCaption = _ref3.subCaption,
        selectionOverlay = _ref3.selectionOverlay,
        rest = _objectWithoutProperties(_ref3, ["caption", "captionComponent", "css", "imageComponent", "placeholder", "source", "subCaption", "selectionOverlay"]);

    delete rest.iconComponent;
    delete rest.selected;
    delete rest.selectionOverlayShowing;
    return _react["default"].createElement(_Layout.Column, Object.assign({}, rest, {
      inline: true
    }), _react["default"].createElement(_Layout.Cell, {
      className: css.image,
      component: ImageComponent,
      placeholder: placeholder,
      src: source
    }, selectionOverlay), caption ? _react["default"].createElement(_Layout.Cell, {
      className: css.caption,
      component: Caption,
      shrink: true
    }, caption) : null, subCaption ? _react["default"].createElement(_Layout.Cell, {
      className: css.subCaption,
      component: Caption,
      shrink: true
    }, subCaption) : null);
  }
});
exports.GridListImageItem = GridListImageItem;
var _default = GridListImageItem;
exports["default"] = _default;
