// @enact/core/internal/prop-types/prop-types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _util = require("../../util");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

var isRequired = function isRequired(fn) {
  fn.isRequired = function (props, key, componentName, location, propFullName) {
    var propValue = props[key];

    if (typeof propValue === 'undefined') {
      return new Error("'".concat(propFullName, "' is required for '").concat(componentName, "' but was undefined."));
    }

    for (var _len = arguments.length, rest = new Array(_len > 5 ? _len - 5 : 0), _key = 5; _key < _len; _key++) {
      rest[_key - 5] = arguments[_key];
    }

    return fn.apply(void 0, [propValue, key, componentName, location, propFullName].concat(rest));
  };

  return fn;
};

var renderable = isRequired(function (props, key, componentName) {
  var propValue = props[key];

  if (propValue && !(0, _util.isRenderable)(propValue)) {
    return new Error("Invalid prop '".concat(key, "' supplied to '").concat(componentName, "'. ") + "Expected a renderable value but received '".concat(_typeof(propValue), "' instead."));
  }
});
var component = isRequired(function (props, key, componentName) {
  var propValue = props[key];

  if (propValue && (typeof propValue === 'string' || !(0, _util.isRenderable)(propValue))) {
    return new Error("Invalid prop '".concat(key, "' supplied to '").concat(componentName, "'. ") + "Expected a function but received '".concat(_typeof(propValue), "' instead."));
  }
});

var renderableOverride = _propTypes["default"].oneOfType([_propTypes["default"].element, renderable]);

var componentOverride = _propTypes["default"].oneOfType([_propTypes["default"].element, component]);

var EnactPropTypes = {
  component: component,
  componentOverride: componentOverride,
  renderable: renderable,
  renderableOverride: renderableOverride
};
var _default = EnactPropTypes;
exports["default"] = _default;
