// @enact/ui/Marquee/MarqueeBase.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MarqueeBase = exports["default"] = void 0;

var _kind = _interopRequireDefault(require("@enact/core/kind"));

var _handle = require("@enact/core/handle");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _MarqueeModule = _interopRequireDefault(require("./Marquee.module.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var isEventSource = function isEventSource(ev) {
  return ev.target === ev.currentTarget;
};
/**
 * Marquees the children of the component.
 *
 * For automated marquee calculations use {@link ui/Marquee.Marquee}.
 *
 * @class MarqueeBase
 * @memberof ui/Marquee
 * @ui
 * @public
 */


var MarqueeBase = (0, _kind["default"])({
  name: 'ui:Marquee',
  propTypes:
  /** @lends ui/Marquee.MarqueeBase.prototype */
  {
    /**
     * Text alignment value of the marquee
     *
     * Valid values are:
     *
     * * `'left'`,
     * * `'right'`, and
     * * `'center'`
     *
     * @type {String}
     * @public
     */
    alignment: _propTypes["default"].oneOf(['left', 'right', 'center']),

    /**
     * Applies animating styles such as removing the ellipsis.
     *
     * @type {Boolean}
     * @public
     */
    animating: _propTypes["default"].bool,

    /**
     * The text or a set of components that should be marqueed
     *
     * This prop may be empty in some cases, which is OK.
     *
     * @type {Node}
     * @public
     */
    children: _propTypes["default"].node,

    /**
     * Called when mounting or unmounting with a reference to the client node
     *
     * @type {Function}
     * @public
     */
    clientRef: _propTypes["default"].func,

    /**
     * Customizes the component by mapping the supplied collection of CSS class names to the
     * corresponding internal Elements and states of this component.
     *
     * The following classes are supported:
     *
     * * `marquee` - The root component class
     * * `animate` - Applied to the inner content node when the text is animating
     * * `text` - The inner content node
     * * `willAnimate` - Applied to the inner content node shortly before animation
     *
     * @type {Object}
     * @public
     */
    css: _propTypes["default"].object,

    /**
     * Distance to animate the marquee
     *
     * Usually, the `distance` is the width of the text minus the width of the container
     *
     * @type {Number}
     * @public
     */
    distance: _propTypes["default"].number,

    /**
     * Called when the marquee completes its animation
     *
     * @type {Function}
     * @public
     */
    onMarqueeComplete: _propTypes["default"].func,

    /**
     * Text overflow setting. Either `'clip'` or `'ellipsis'`
     *
     * @type {String}
     * @public
     */
    overflow: _propTypes["default"].oneOf(['clip', 'ellipsis']),

    /**
     * `true` if the directionality of the content is right-to-left
     *
     * @type {Boolean}
     * @public
     * @default false
     */
    rtl: _propTypes["default"].bool,

    /**
     * Speed of marquee animation in pixels/second.
     *
     * @type {Number}
     * @public
     */
    speed: _propTypes["default"].number,

    /**
     * Indicates the marquee will animate soon.
     *
     * Should be used by the component to prepare itself for animation such as promoting
     * composite layers for improved performance.
     *
     * @type {Boolean}
     * @public
     * @default false
     */
    willAnimate: _propTypes["default"].bool
  },
  defaultProps: {
    rtl: false,
    willAnimate: false
  },
  styles: {
    css: _MarqueeModule["default"],
    className: 'marquee',
    publicClassNames: true
  },
  handlers: {
    onMarqueeComplete: (0, _handle.handle)((0, _handle.forProp)('animating', true), isEventSource, _handle.stop, function (ev, props) {
      return (0, _handle.forward)('onMarqueeComplete', {
        type: 'onMarqueeComplete'
      }, props);
    })
  },
  computed: {
    clientClassName: function clientClassName(_ref) {
      var animating = _ref.animating,
          willAnimate = _ref.willAnimate,
          styler = _ref.styler;
      return styler.join({
        animate: animating,
        text: true,
        willAnimate: willAnimate
      });
    },
    clientStyle: function clientStyle(_ref2) {
      var alignment = _ref2.alignment,
          animating = _ref2.animating,
          distance = _ref2.distance,
          overflow = _ref2.overflow,
          rtl = _ref2.rtl,
          speed = _ref2.speed;
      // If the components content directionality doesn't match the context, we need to set it
      // inline
      var direction = rtl ? 'rtl' : 'ltr';
      var sideProperty = rtl ? 'left' : 'right';
      var style = {
        direction: direction,
        textAlign: alignment,
        textOverflow: overflow
      };

      if (animating) {
        var duration = distance / speed;
        style[sideProperty] = "".concat(distance, "px");
        style.transitionDuration = "".concat(duration, "s");
      } else {
        style[sideProperty] = 0;
      }

      return style;
    }
  },
  render: function render(_ref3) {
    var children = _ref3.children,
        clientClassName = _ref3.clientClassName,
        clientRef = _ref3.clientRef,
        clientStyle = _ref3.clientStyle,
        onMarqueeComplete = _ref3.onMarqueeComplete,
        rest = _objectWithoutProperties(_ref3, ["children", "clientClassName", "clientRef", "clientStyle", "onMarqueeComplete"]);

    delete rest.alignment;
    delete rest.animating;
    delete rest.distance;
    delete rest.onMarqueeComplete;
    delete rest.overflow;
    delete rest.rtl;
    delete rest.speed;
    delete rest.willAnimate;
    return _react["default"].createElement("div", rest, _react["default"].createElement("div", {
      className: clientClassName,
      ref: clientRef,
      style: clientStyle,
      onTransitionEnd: onMarqueeComplete
    }, children));
  }
});
exports.MarqueeBase = MarqueeBase;
var _default = MarqueeBase;
exports["default"] = _default;
