// @enact/ui/Image/Image.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ImageBase = exports.Image = exports["default"] = void 0;

var _kind = _interopRequireDefault(require("@enact/core/kind"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _warning = _interopRequireDefault(require("warning"));

var _resolution = require("../resolution");

var _ImageModule = _interopRequireDefault(require("./Image.module.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/**
 * A basic image component designed to display images conditionally based on screen size.
 *
 * This component does not have a default size, therefore the image will not show unless a size is
 * specified using a CSS `className` or inline `style`.
 *
 * Example:
 * ```
 * const src = {
 *   'hd': 'http://lorempixel.com/64/64/city/1/',
 *   'fhd': 'http://lorempixel.com/128/128/city/1/',
 *   'uhd': 'http://lorempixel.com/256/256/city/1/'
 * };
 *
 * <Image className={css.myImage} src={src} sizing="fill" />
 * ```
 *
 * `ui/Image` is based on the `div` element, but it uses `img` to provide `onError` and `onLoad`
 * events. The image that you see on screen is a `background-image` from the `div` element, not the
 * `img` element.
 *
 * > If you need a naturally sized image, you can use the native `<img>` element instead.
 *
 * @class Image
 * @memberof ui/Image
 * @ui
 * @public
 */
var ImageBase = (0, _kind["default"])({
  name: 'ui:Image',
  propTypes:
  /** @lends ui/Image.Image.prototype */
  {
    /**
     * String value for the alt attribute of the image.
     *
     * @type {String}
     * @public
     */
    alt: _propTypes["default"].string,

    /**
     * The aria-label for the image.
     *
     * If unset, it defaults to the value of `alt`.
     *
     * @type {String}
     * @public
     * @memberof ui/Image.Image.prototype
     */
    'aria-label': _propTypes["default"].string,

    /**
     * Node for the children of an `Image`. Useful for overlays.
     *
     * @type {Node}
     * @public
     */
    children: _propTypes["default"].node,

    /**
     * Customizes the component by mapping the supplied collection of CSS class names to the
     * corresponding internal Elements and states of this component.
     *
     * The following classes are supported:
     *
     * * `image` - The root component class
     * * `fit` - Applied when `sizing` prop is `fit`
     * * `fill` - Applied when `sizing` prop is `fill`
     *
     * @type {Object}
     * @public
     */
    css: _propTypes["default"].object,

    /**
     * Called if the image has an error.
     *
     * @type {Function}
     * @public
     */
    onError: _propTypes["default"].func,

    /**
     * Called once the image is loaded.
     *
     * @type {Function}
     * @public
     */
    onLoad: _propTypes["default"].func,

    /**
     * A placeholder image to be displayed before the image is loaded.
     *
     * For performance purposes, it should be pre-loaded or be a data url. If `src` is
     * unset, this value will be used as the `background-image`.
     *
     * @type {String}
     * @default ''
     * @public
     */
    placeholder: _propTypes["default"].string,

    /**
     * Used to set the `background-size` of an `Image`.
     *
     * * `'fill'` - sets `background-size: cover`
     * * `'fit'` - sets `background-size: contain`
     * * `'none'` - uses the image's natural size
     *
     * @type {String}
     * @default 'fill'
     * @public
     */
    sizing: _propTypes["default"].oneOf(['fit', 'fill', 'none']),

    /**
     * String value or Object of values used to determine which image will appear on
     * a specific screenSize.
     *
     * @type {String|Object}
     * @public
     */
    src: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].object])
  },
  defaultProps: {
    placeholder: '',
    sizing: 'fill'
  },
  styles: {
    css: _ImageModule["default"],
    className: 'image',
    publicClassNames: true
  },
  computed: {
    bgImage: function bgImage(_ref) {
      var placeholder = _ref.placeholder,
          src = _ref.src;
      var imageSrc = (0, _resolution.selectSrc)(src) || placeholder;
      process.env.NODE_ENV !== "production" ? (0, _warning["default"])(imageSrc, 'Image requires that either the "placeholder" or "src" props be specified.') : void 0;
      if (!imageSrc) return null;
      return placeholder && imageSrc !== placeholder ? "url(\"".concat(imageSrc, "\"), url(\"").concat(placeholder, "\")") : "url(\"".concat(imageSrc, "\")");
    },
    className: function className(_ref2) {
      var _className = _ref2.className,
          sizing = _ref2.sizing,
          styler = _ref2.styler;
      return sizing !== 'none' ? styler.append(sizing) : _className;
    },
    imgSrc: function imgSrc(_ref3) {
      var src = _ref3.src;
      return (0, _resolution.selectSrc)(src) || null;
    }
  },
  render: function render(_ref4) {
    var alt = _ref4.alt,
        ariaLabel = _ref4['aria-label'],
        bgImage = _ref4.bgImage,
        children = _ref4.children,
        css = _ref4.css,
        imgSrc = _ref4.imgSrc,
        onError = _ref4.onError,
        onLoad = _ref4.onLoad,
        style = _ref4.style,
        rest = _objectWithoutProperties(_ref4, ["alt", "aria-label", "bgImage", "children", "css", "imgSrc", "onError", "onLoad", "style"]);

    delete rest.placeholder;
    delete rest.sizing;
    delete rest.src;
    return _react["default"].createElement("div", Object.assign({
      role: "img"
    }, rest, {
      "aria-label": ariaLabel || alt,
      style: _objectSpread({}, style, {
        backgroundImage: bgImage
      })
    }), children, _react["default"].createElement("img", {
      className: css.img,
      src: imgSrc,
      alt: alt,
      onLoad: onLoad,
      onError: onError
    }));
  }
});
exports.ImageBase = exports.Image = ImageBase;
var _default = ImageBase;
exports["default"] = _default;
