// @enact/ui/ProgressBar/ProgressBar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ProgressBar = exports["default"] = void 0;

var _kind = _interopRequireDefault(require("@enact/core/kind"));

var _clamp = _interopRequireDefault(require("ramda/src/clamp"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _validators = require("../internal/validators");

var _ProgressBarModule = _interopRequireDefault(require("./ProgressBar.module.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var progressToProportion = function progressToProportion(value) {
  return (0, _clamp["default"])(0, 1, value);
};

var calcBarStyle = function calcBarStyle(prop, anchor) {
  var _ref;

  var value = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : anchor;
  var startProp = arguments.length > 3 ? arguments[3] : undefined;
  var endProp = arguments.length > 4 ? arguments[4] : undefined;
  var start = Math.min(anchor, value);
  var end = Math.max(anchor, value) - start;

  if (process.env.NODE_ENV !== "production") {
    (0, _validators.validateRange)(start, 0, 1, 'ProgressBar', prop, 'min', 'max');
    (0, _validators.validateRange)(end, 0, 1, 'ProgressBar', prop, 'min', 'max');
  }

  return _ref = {}, _defineProperty(_ref, startProp, progressToProportion(start)), _defineProperty(_ref, endProp, progressToProportion(end)), _ref;
};
/**
 * An unstyled progress bar component that can be customized by a theme or application.
 *
 * @class ProgressBar
 * @memberof ui/ProgressBar
 * @ui
 * @public
 */


var ProgressBar = (0, _kind["default"])({
  name: 'ui:ProgressBar',
  propTypes:
  /** @lends ui/ProgressBar.ProgressBar.prototype */
  {
    /**
     * The proportion of the loaded portion of the progress bar.
     *
     * * Valid values are between `0` and `1`.
     *
     * @type {Number}
     * @default 0
     * @public
     */
    backgroundProgress: _propTypes["default"].number,

    /**
     * The contents to be displayed with progress bar.
     *
     * @type {Node}
     * @public
     */
    children: _propTypes["default"].node,

    /**
     * Customizes the component by mapping the supplied collection of CSS class names to the
     * corresponding internal Elements and states of this component.
     *
     * The following classes are supported:
     *
     * * `progressBar` - The root component class
     * * `fill`        - The foreground node of the progress bar
     * * `load`        - The background node of the progress bar
     * * `horizontal`  - Applied when `orientation` is `'horizontal'`
     * * `vertical`    - Applied when `orientation` is `'vertical'`
     *
     * @type {Object}
     * @public
     */
    css: _propTypes["default"].object,

    /**
     * Sets the orientation of the slider.
     *
     * Allowed values include:
     *
     * * `'horizontal'` - A left and right orientation
     * * `'vertical'` - An up and down orientation
     *
     * @type {String}
     * @default 'horizontal'
     * @public
     */
    orientation: _propTypes["default"].string,

    /**
     * The proportion of the filled portion of the progress bar.
     *
     * * Valid values are between `0` and `1`.
     *
     * @type {Number}
     * @default 0
     * @public
     */
    progress: _propTypes["default"].number,

    /**
     * Sets the point, as a proportion between 0 and 1, from which the progress bar is filled.
     *
     * Applies to both the progress bar's `value` and `backgroundProgress`. In both cases,
     * setting the value of `progressAnchor` will cause the progress bar to fill from that point
     * down, when `value` or `backgroundProgress` is proportionally less than the anchor, or up,
     * when `value` or `backgroundProgress` is proportionally greater than the anchor, rather
     * than always from the start of the progress bar.
     *
     * @type {Number}
     * @default 0
     * @public
     */
    progressAnchor: _propTypes["default"].number
  },
  defaultProps: {
    backgroundProgress: 0,
    orientation: 'horizontal',
    progress: 0,
    progressAnchor: 0
  },
  styles: {
    css: _ProgressBarModule["default"],
    className: 'progressBar',
    publicClassNames: true
  },
  computed: {
    className: function className(_ref2) {
      var orientation = _ref2.orientation,
          styler = _ref2.styler;
      return styler.append(orientation);
    },
    style: function style(_ref3) {
      var backgroundProgress = _ref3.backgroundProgress,
          progress = _ref3.progress,
          progressAnchor = _ref3.progressAnchor,
          _style = _ref3.style;
      return _objectSpread({}, _style, calcBarStyle('backgroundProgress', progressAnchor, backgroundProgress, '--ui-progressbar-proportion-start-background', '--ui-progressbar-proportion-end-background'), calcBarStyle('progress', progressAnchor, progress, '--ui-progressbar-proportion-start', '--ui-progressbar-proportion-end'));
    }
  },
  render: function render(_ref4) {
    var children = _ref4.children,
        css = _ref4.css,
        rest = _objectWithoutProperties(_ref4, ["children", "css"]);

    delete rest.backgroundProgress;
    delete rest.orientation;
    delete rest.progress;
    delete rest.progressAnchor;
    return _react["default"].createElement("div", Object.assign({
      role: "progressbar"
    }, rest), _react["default"].createElement("div", {
      className: css.load
    }), _react["default"].createElement("div", {
      className: css.fill
    }), children);
  }
});
exports.ProgressBar = ProgressBar;
var _default = ProgressBar;
exports["default"] = _default;
