// @enact/ui/Scrollable/ScrollThumb.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ScrollThumbBase = exports.ScrollThumb = exports["default"] = void 0;

var _kind = _interopRequireDefault(require("@enact/core/kind"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _ScrollThumbModule = _interopRequireDefault(require("./ScrollThumb.module.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

/**
 * An unstyled scroll thumb without any behavior.
 *
 * @class ScrollThumb
 * @memberof ui/Scrollable
 * @ui
 * @private
 */
var ScrollThumb = (0, _kind["default"])({
  name: 'ui:ScrollThumb',
  propTypes:
  /** @lends ui/Scrollable.ScrollThumb.prototype */
  {
    /**
     * If `true`, the scrollbar will be oriented vertically.
     *
     * @type {Boolean}
     * @default true
     * @public
     */
    vertical: _propTypes["default"].bool
  },
  defaultProps: {
    vertical: true
  },
  styles: {
    css: _ScrollThumbModule["default"],
    className: 'scrollThumb'
  },
  computed: {
    className: function className(_ref) {
      var vertical = _ref.vertical,
          styler = _ref.styler;
      return styler.append({
        vertical: vertical
      });
    }
  },
  render: function render(props) {
    delete props.vertical;
    return _react["default"].createElement("div", props);
  }
});
exports.ScrollThumbBase = exports.ScrollThumb = ScrollThumb;
var _default = ScrollThumb;
exports["default"] = _default;
